Fix building 32bit Flute

ivanmgribeiro opened the pull request:
I'm not sure whether using `XLEN_2` is the right thing to do, or whether it would be better to use something else (like `CapW` or `TSub #(CapMem, 1)`)

ivanmgribeiro committed:
Fix building 32bit Flute

jrtc27 commented:
The whole point of introducing CWord was so that it could be made 64-bit on RV32 again like upstream, with the growth to 128-bit only needed on RV64 to accommodate 128-bit capabilities.

jrtc27 commented on src_Core/CPU/CPU_Stage2.bsv:
```diff
@@ -449,11 +449,12 @@ module mkCPU_Stage2 #(Bit #(4)         verbosity,
                                                  rg_f5,
                                                  rg_stage2.addr & ~(~0 << rg_stage2.mem_width_code), // force aligned address as return from memory is already sliced appropriately (a 0 addr would just work)
-                                                 unpack(pack(toMem(result))),
+                                                 unpack({pack (res_tag), zeroExtend (res_data)}),
```
If CWord is staying as 128-bit for now then this is wrong for RV32, and highlights why unpack(pack()) is a bad idea when combined with extension or truncation, and you should instead be doing:
```
{res_tag, zeroExtend (res_data)}
```
because you really want to make sure that tags and data never spill over into each other. As it stands, I assume Cache_Cap_Tag_Width is 2 for RV32 due to the 128-bit cache lines, and so half the time the tag is in the right place and half the time it's not and ends up being a 0 from the zeroExtend spilling into the tags.

ivanmgribeiro replied:
Thanks for pointing this out. I've pushed a commit that fixes this.
It also took me a long time to realise this, but it looks like this code is only compiled in RVFI and is only used to find the write data mask, so I don't think this actually matters here. However, seeing this did lead to finding a bug in the caches where the tag for the lower word was not being set.

jrtc27 commented on src_Core/Near_Mem_VM_WT_L1/MMU_Cache.bsv:
```diff
@@ -105,7 +105,7 @@ interface MMU_Cache_IFC #(numeric type mID);
 		       Bit #(5) amo_funct5,
 `endif
 		       Addr addr,
-               Tuple2#(Bool, Bit#(128)) st_value,
```
This should be CWord and propagated throughout (see the upstream source that is now using it)

ivanmgribeiro committed:
Fix bug pointed out by @jrtc27

ivanmgribeiro committed:
Don't ignore lower word64 user bits

ivanmgribeiro committed:
Change caches from 128bit to 64bit on RV32

ivanmgribeiro committed:
Clean up cache refill loop

ivanmgribeiro committed:
Add error message & comments on RV32 multi-flit IO

ivanmgribeiro committed:
Update WB_L1_L2 caches to support RV32 and CHERI64

The LLC Cache is currently hardwired to 512b cache lines. In order to
maintain compatibility with this, the RV32 CHERI cache definitions are
changed to have 512b cache lines (ie 8*64b rather than 4*64b).

ivanmgribeiro committed:
Add changes suggested by @gameboo

ivanmgribeiro committed:
Fix commit handling in WB_L1_L2

jrtc27 commented:
Can you please squash your commits together next time rather than creating a bunch of fixups that clutter the history?
